lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1760,6 +1760,38 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
   case OpGetKernelNDrangeSubGroupCount:
     return mapValue(
         BV, transSGSizeQueryBI(static_cast<SPIRVInstruction *>(BV), BB));
+  case OpFPGARegINTEL: {
+    IRBuilder<> Builder(BB);
+
+    SPIRVFPGARegINTELInstBase *BC =
+        static_cast<SPIRVFPGARegINTELInstBase *>(BV);
+
+    PointerType *Int8PtrTyPrivate =
+        Type::getInt8PtrTy(*Context, SPIRAS_Private);
+    IntegerType *Int32Ty = Type::getInt32Ty(*Context);
+
+    Value *UndefInt8Ptr = UndefValue::get(Int8PtrTyPrivate);
+    Value *UndefInt32 = UndefValue::get(Int32Ty);
+
+    Constant *GS = Builder.CreateGlobalStringPtr(kOCLBuiltinName::FPGARegIntel);
+
+    Type *Ty = transType(BC->getType());
+    llvm::Value *Args[] = {transValue(BC->getOperand(0), F, BB), GS,
+                           UndefInt8Ptr, UndefInt32};
+
+    CallInst *IntrinsicCall = nullptr;
+    if (auto *RetIntTy = dyn_cast<IntegerType>(Ty))
+      IntrinsicCall =
+          Builder.CreateIntrinsic(Intrinsic::annotation, RetIntTy, Args);
+    else if (auto *RetPtrTy = dyn_cast<PointerType>(Ty))
+      if (isa<IntegerType>(RetPtrTy->getElementType()))
+        IntrinsicCall =
+            Builder.CreateIntrinsic(Intrinsic::ptr_annotation, RetPtrTy, Args);
+    assert(IntrinsicCall &&
```

Comment:
According to the extension spec, OpFPGARegINTEL can have operands of any type. This implementation is limited to integer and pointer to integer types.
Should we insert a bitcast for other types here (to make it consumable by llvm intrinsics) ?

Reply:
For the current scenarios, no other type is going to arrive to SPIR-V, as Translator creates FPGARegIntel from the same LLVM intrinsics anyway. However, I agree I should bitcast here to lessen dependency on LLVM->SPIR-V translation module and achieve spec compliance.